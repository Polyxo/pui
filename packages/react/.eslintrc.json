{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint"],
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:storybook/recommended"
  ],
  "rules": {
    "no-unused-vars": "off",
    "no-console": "error",
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
