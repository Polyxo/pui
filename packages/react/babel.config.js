module.exports = {
  presets: [
    "@babel/preset-env",
    "@babel/preset-react",
    "@babel/preset-typescript",
  ],
  plugins: [
    "dev-expression",
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-syntax-import-meta",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-export-namespace-from",
    "@babel/plugin-proposal-export-default-from",
  ],
};
