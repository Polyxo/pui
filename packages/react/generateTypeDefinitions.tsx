import { rm, stat, readdir, writeFile, mkdir } from 'fs/promises';
import path from 'path';
import { parse } from 'react-docgen-typescript';

const distUrl = path.resolve(__dirname, '../../website/types/');

const options = {
  savePropValueAsString: true,
  propFilter: (prop, component) => {
    if (prop.declarations !== undefined && prop.declarations.length > 0) {
      const hasPropAdditionalDescription = prop.declarations.find(
        (declaration) => {
          return !declaration.fileName.includes('node_modules');
        }
      );

      return Boolean(hasPropAdditionalDescription);
    }

    return true;
  },
};

export default async function walk(directory: string) {
  let fileList: string[] = [];

  const files = await readdir(directory);
  for (const file of files) {
    const p = path.join(directory, file);
    if ((await stat(p)).isDirectory()) {
      fileList = [...fileList, ...(await walk(p))];
    } else {
      fileList.push(p);
    }
  }

  const tsxFileList = fileList.filter(
    (file) => file.endsWith('.tsx') && !file.includes('stories')
  );

  await Promise.all(
    tsxFileList.map(async (file) => {
      const propTypes = JSON.parse(JSON.stringify(parse(file, options)));

      try {
        await mkdir(path.dirname(path.join(distUrl, file)), {
          recursive: true,
        });
      } catch (error) {
        console.log('collectTypes error', error);
      }

      console.log(
        'propTypes',
        path.join(distUrl, file.replace('.tsx', '.json'))
      );
      await writeFile(
        path.join(distUrl, file.replace('.tsx', '.json')),
        JSON.stringify(propTypes, null, 2)
      );
      return null;
    })
  );

  return fileList;
}

async function startGenerateTypes() {
  await rm(distUrl, { recursive: true, force: true });

  await walk('./src');
  process.exit();
}

startGenerateTypes();
