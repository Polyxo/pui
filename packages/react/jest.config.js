module.exports = {
  preset: "ts-jest",
  testEnvironment: "jsdom",
  transform: {
    "^.+\\.tsx?$": "ts-jest",
  },

  roots: ["<rootDir>/src"],
  testMatch: [
    //"**/__tests__/**/*.+(ts|tsx|js)",
    "**/?(*.)+(spec|test).+(ts|tsx|js)",
  ],
};
