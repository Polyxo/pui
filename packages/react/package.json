{
  "name": "@wfp/react",
  "version": "1.19.3",
  "description": "WFP UI Kit",
  "license": "Apache-2",
  "main": "lib/index.js",
  "module": "es/index.js",
  "main:scss": "source/globals/styles.scss",
  "types": "es/dist/index.d.ts",
  "scripts": {
    "build": "yarn rollup -c --bundleConfigAsCjs && yarn build:demos && yarn generate:types",
    "clean": "rimraf es lib umd storybook-static build react-docgen.json",
    "build:rollup": "rollup -c --bundleConfigAsCjs",
    "build:demos": "rollup -c rollupDemos.config.js --bundleConfigAsCjs",
    "build:storybook": "rm -r -f docs && storybook build -o docs",
    "generate:types": "rollup --config generateTypeDefinitions/rollup.config.js --bundleConfigAsCjs && node .temp/generateTypeDefinitions.js && rm -rf .temp",
    "dev": "npm run storybook",
    "storybook": "STORYBOOK_ASSETS=https://cdn.wfp.org/guides/ui/assets/v0.0.1/ STORYBOOK_NPM_VERSION=10.10.3 storybook dev -p 9000",
    "test": "jest",
    "jest-w": "jest --watch"
  },
  "sass": {
    "includePaths": [
      "./node_modules"
    ]
  },
  "keywords": [
    "WFP",
    "World Food Programme",
    "UN",
    "United Nations",
    "react",
    "ui-kit",
    "components"
  ],
  "babel": {
    "presets": [
      "@babel/preset-react"
    ],
    "plugins": [
      "dev-expression",
      "@babel/plugin-syntax-dynamic-import",
      "@babel/plugin-syntax-import-meta",
      "@babel/plugin-proposal-class-properties",
      "@babel/plugin-proposal-export-namespace-from",
      "@babel/plugin-proposal-export-default-from"
    ]
  },
  "files": [
    ".storybook",
    "build",
    "config",
    "es",
    "lib",
    "src",
    "umd",
    "rollup.config.ts",
    "react-docgen.json"
  ],
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "dependencies": {
    "@babel/preset-typescript": "^7.18.6",
    "@carbon/feature-flags": "^0.6.0",
    "classnames": "2.2.6",
    "invariant": "^2.2.4",
    "react-datepicker": "^4.6.0",
    "react-popper-tooltip": "^4.1.0",
    "react-transition-state": "^2.1.0",
    "warning": "^4.0.3",
    "window-or-global": "^1.0.1",
    "ws": "^8.12.0"
  },
  "devDependencies": {
    "@babel/traverse": "^7.23.2",
    "@fontsource-variable/open-sans": "^5.0.13",
    "@rollup/plugin-babel": "^6.0.4",
    "@rollup/plugin-commonjs": "^25.0.7",
    "@rollup/plugin-node-resolve": "^15.2.3",
    "@rollup/plugin-replace": "^5.0.2",
    "@rollup/plugin-terser": "^0.4.3",
    "@rollup/plugin-typescript": "^11.1.5",
    "@storybook/addon-docs": "^7.5.3",
    "@storybook/addon-essentials": "^7.5.3",
    "@storybook/addon-links": "^7.5.3",
    "@storybook/addon-mdx-gfm": "^7.5.3",
    "@storybook/addon-styling": "^1.3.7",
    "@storybook/blocks": "^7.5.3",
    "@storybook/react": "^7.5.3",
    "@storybook/react-vite": "^7.5.3",
    "@storybook/testing-library": "^0.2.2",
    "@types/jest": "^29.0.3",
    "@types/mdx": "^2.0.3",
    "@types/node": "^18.7.18",
    "@types/react": "^18.2.14",
    "@types/react-dom": "^18.0.6",
    "@typescript-eslint/eslint-plugin": "^5.56.0",
    "@typescript-eslint/parser": "^5.56.0",
    "@un/babel-preset-wfp": "^0.2.3",
    "@un/icons-react": "^10.35.3",
    "@un/pictograms-react": "^11.15.3",
    "@wfp/styles": "^10.27.3",
    "@wfp/themes-core": "^0.0.48",
    "all-contributors-cli": "^6.17.0",
    "autoprefixer": "^9.8.6",
    "babel-eslint": "^10.1.0",
    "babel-loader": "^8.2.2",
    "babel-plugin-dev-expression": "^0.2.2",
    "babel-plugin-named-exports-order": "^0.0.2",
    "browserslist-config-carbon": "^10.6.1",
    "clean-webpack-plugin": "^3.0.0",
    "css-loader": "^5.2.4",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.2",
    "enzyme-to-json": "^3.5.0",
    "eslint-plugin-storybook": "^0.6.15",
    "gzip-size": "^5.1.1",
    "handlebars": "^4.7.7",
    "iframe-resizer": "^4.3.6",
    "jest": "^26.2.2",
    "lodash.isequal": "^4.5.0",
    "moment": "^2.27.0",
    "object-assign": "^4.1.1",
    "postcss": "^8.2.14",
    "postcss-cli": "^8.3.1",
    "postcss-custom-properties": "^11.0.0",
    "postcss-loader": "4",
    "prettier": "^2.8.7",
    "pretty": "^2.0.0",
    "prismjs": "^1.20.0",
    "promise": "^8.1.0",
    "prop-types": "^15.8.1",
    "raw-loader": "^4.0.1",
    "react": "18.2.0",
    "react-dates": "^21.8.0",
    "react-docgen": "^7.0.0",
    "react-docgen-typescript": "^2.2.2",
    "react-dom": "18.2.0",
    "react-dropzone": "^11.0.2",
    "react-final-form": "^6.5.1",
    "react-hook-form": "^7.20.5",
    "react-markdown": "^9.0.0",
    "react-popper": "^2.2.3",
    "react-prism": "^4.3.2",
    "react-redux": "^7.2.1",
    "react-select": "^3.1.0",
    "react-table": "7.5.0",
    "recharts": "^2.7.3",
    "redux-form": "^8.3.10",
    "requestanimationframe": "^0.0.23",
    "rimraf": "^3.0.2",
    "rollup": "4.1.4",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-preserve-directives": "^0.2.0",
    "rollup-plugin-string": "^3.0.0",
    "rollup-plugin-strip-banner": "^3.0.0",
    "rollup-plugin-terser": "^7.0.2",
    "sass": "^1.66.1",
    "sass-loader": "13.3.2",
    "storybook": "^7.5.3",
    "style-loader": "^2.0.0",
    "twig": "^1.15.4",
    "typescript": "^4.8.3",
    "vite": "^4.4.9"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/wfp/ui.git"
  },
  "bundleSizeThreshold": 80000,
  "gitHead": "5d9678f4a62413b834d07736413c50ff2f851824"
}
