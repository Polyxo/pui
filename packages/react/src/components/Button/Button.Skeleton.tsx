import * as React from "react";
import classNames from "classnames";

interface ButtonSkeletonProps {
  small?: boolean;
  href?: string;
}

const ButtonSkeleton: React.FC<ButtonSkeletonProps> = ({
  small = false,
  href,
}) => {
  const buttonClasses = classNames({
    "wfp--skeleton": true,
    "wfp--btn": true,
    "wfp--btn--sm": small,
  });

  const commonProps = {
    className: buttonClasses,
  };

  const button = <button {...commonProps} type="button" />;

  const anchor = <a {...commonProps} href={href} role="button" />; // eslint-disable-line

  return href ? anchor : button;
};

export default ButtonSkeleton;
