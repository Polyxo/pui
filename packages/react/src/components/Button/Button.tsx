import * as React from "react";
import { useState, useEffect } from "react";

import classNames from "classnames";
import useSettings from "../../hooks/useSettings";
import { ButtonKind } from "../../utils";

interface ButtonBaseProps {
  /**
   * Content of the Button @design
   */
  children?: React.ReactNode;
  /**
   * Specify whether the Button should be a small variant @design
   */
  small?: boolean;
  /**
   * Specify whether the Button should be a large variant
   */
  large?: boolean;

  /**
   * Specify btnSolid to buttons on solid background to make it stand out
   */
  btnSolid?: boolean;
  /**
   * Specify the kind of Button you want to create @design
   */
  kind?: ButtonKind;
  /**
   * Optionally specify an href for your Button to become an element
   */
  href?: string;
  //target?: string; // TODO: Remove  " "
  /**
   * Specify an `icon` to include in the Button through an object representing the SVG data of the icon, similar to the `Icon` component @design
   */
  icon?: React.ReactNode | React.ComponentType<any>;
  /**
   * Optionally specify an href for your Button to become an element @design
   */
  iconReverse?: boolean;
  // todo: revise iconDescription
  iconDescription?: string;
  /**
   * Specify whether the Button should be disabled, or not
   */
  disabled?: boolean;
  /**
   * Optional prop to specify the tabIndex of the Button
   */
  tabIndex?: number;
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>["target"];
  // type?: 'button' | 'reset' | 'submit';
  useFlexbox?: boolean;
  id?: string;
}

interface ButtonButtonProps
  extends ButtonBaseProps,
    React.ComponentPropsWithoutRef<"button"> {}

interface ButtonLinkProps
  extends ButtonBaseProps,
    React.ComponentPropsWithoutRef<"a"> {}

type ConditionalProps<T> = T extends { href: string }
  ? ButtonLinkProps
  : ButtonButtonProps;

/**
 * Buttons express what action will occur when the user clicks or touches it. Buttons are used to initialize an action, either in the background or foreground of an experience. */

export type ButtonRef<T extends React.ElementType> =
  React.ComponentPropsWithRef<T>["ref"];

export const Button = React.forwardRef(
  <T extends { href?: string }>(
    props: ConditionalProps<T>,
    ref: React.Ref<
      T extends { href: string } ? HTMLLinkElement : HTMLButtonElement
    >
  ) => {
    const {
      children,
      className,
      disabled,
      small,
      large,
      btnSolid,
      kind = "primary",
      iconReverse,
      tabIndex,
      useFlexbox,
      type = "button",
      icon,
      iconDescription,
      onClick,
      id,
      ...other
    } = props;
    const { prefix } = useSettings();
    const [count, setCount] = useState(false);

    useEffect(() => {
      const timer = setTimeout(() => endAnimation(), 500);
      return () => {
        clearTimeout(timer);
      };
    }, [count]);

    const buttonClasses = classNames(className, {
      [`${prefix}--btn`]: true,
      [`${prefix}--btn--sm`]: small,
      [`${prefix}--btn--lg`]: large,
      [`${prefix}--btn--icon-reverse`]: iconReverse,
      [`${prefix}--btn--flexbox`]: useFlexbox,
      [`${prefix}--btn--icon-only`]: icon && children === undefined,
      [`${prefix}--btn--${kind}`]: kind,
      [`${prefix}--btn--${kind}-solid`]: btnSolid,
      [`${prefix}--btn--animating`]: count,
    });

    const commonProps = {
      tabIndex,
      className: buttonClasses,
    };

    let buttonImage: React.ReactNode = null;

    if (icon && React.isValidElement(icon)) {
      buttonImage = <span className={`${prefix}--btn__icon`}>{icon}</span>;
    } else if (icon) {
      const Icon = icon as any;
      buttonImage = (
        <Icon
          description={iconDescription}
          className={`${prefix}--btn__icon`}
        />
      );
    }

    const endAnimation = () => {
      setCount(false);
    };

    const onClickAnimation = (e) => {
      if (onClick) {
        onClick(e);
      }
      setCount(true);
    };

    if (other.href) {
      const anchorProps = {
        ...other,
        href: other.href,
      } as React.AnchorHTMLAttributes<HTMLAnchorElement>;

      return (
        <a
          {...anchorProps}
          {...commonProps}
          role="button"
          onClick={onClickAnimation}
          ref={ref as React.Ref<HTMLAnchorElement>}
          id={id}
        >
          {iconReverse && buttonImage}
          {children}
          {!iconReverse && buttonImage}
        </a>
      );
    }

    const buttonProps = {
      ...other,
      disabled: disabled,
    } as React.ButtonHTMLAttributes<HTMLButtonElement>;

    return (
      <button
        {...buttonProps}
        {...commonProps}
        disabled={disabled}
        type={type}
        onClick={onClickAnimation}
        ref={ref as React.Ref<HTMLButtonElement>}
        id={id}
      >
        {iconReverse && buttonImage}
        {children}
        {!iconReverse && buttonImage}
      </button>
    );
  }
);

Button.displayName = "Button";

export default Button;
