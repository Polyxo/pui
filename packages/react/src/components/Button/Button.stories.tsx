import React from "react";
import type { Meta, StoryObj } from "@storybook/react";
import { buttonKinds } from "../../prop-types/types";
import { AddCircle } from "@un/icons-react";
import markdown from "./README.mdx";
import Button from "./Button";

const meta: Meta<typeof Button> = {
  title: "Components/UI Elements/Button",
  component: Button,
  tags: ["autodocs"],
  parameters: {
    componentSubtitle: "Component",
    status: "released",
    mdx: markdown,
    // twig: buttonTwig,
  },
};

// export default {
//   title: "Components/UI Elements/Button",
//   component: Button,
//   markdown: "hello",
//   parameters: {
//     componentSubtitle: "Component",
//     status: "released",
//     mdx: markdown,
//     // twig: buttonTwig,
//   },
// };

export default meta;
type Story = StoryObj<typeof Button>;

//export const ButtonDefault = (args) => <Button {...args}>Default</Button>;

export const ButtonDefault: Story = {
  render: (args) => <Button {...args} />,
  args: {
    children: "Button",
  },
  parameters: {
    design: {
      type: "figma",
      url: "https://www.figma.com/file/jtKqyyMTOKgUTstiGvPZMl/?node-id=0%3A1",
    },
  },
};

export const PrimaryButton = (args) => (
  <>
    <Button {...args}>Default</Button> {""}
    <Button {...args} icon={AddCircle} iconReverse>
      With Icon
    </Button>{" "}
    <Button {...args} disabled>
      Disable
    </Button>
  </>
);

PrimaryButton.args = {
  kind: "primary",
};

export const SecondaryButton = (args) => (
  <>
    <Button {...args}>Default</Button> {""}
    <Button {...args} icon={AddCircle}>
      With Icon
    </Button>{" "}
    <Button {...args} disabled>
      Disable
    </Button>
  </>
);

SecondaryButton.args = {
  kind: "secondary",
};

export const GhostButton = (args) => (
  <>
    <Button {...args}>Default</Button> {""}
    <Button {...args} icon={AddCircle}>
      With Icon
    </Button>{" "}
    <Button {...args} disabled>
      Disable
    </Button>
  </>
);

GhostButton.args = {
  kind: "ghost",
};

export const ButtonSolid = (args) => (
  <div
    style={{
      width: "100%",
      height: "150px",
      backgroundColor: "#017DBC",
      padding: "20px",
    }}
  >
    {buttonKinds
      .filter((e) => e === "primary" || e === "secondary" || e === "ghost")
      .map((e) => (
        <div
          key={e}
          style={{
            display: "inline-block",
            padding: "0.3rem",
          }}
        >
          <Button {...args} kind={e} icon={AddCircle}>
            {e}
          </Button>{" "}
          <Button {...args} disabled kind={e}>
            {e} disable
          </Button>
        </div>
      ))}
  </div>
);

ButtonSolid.args = {
  kind: "primary",
  btnSolid: true,
};

ButtonSolid.parameters = {
  code: false,
};

const solidDesc =
  "This is how the different kinds of buttons look on solid background. Set`solid prop` to true to enable it";

ButtonSolid.parameters = {
  docs: {
    storyDescription: solidDesc,
  },
};

// const hello = `
// By changing the \`kind\` prop you can use different styles of the button.

// | kind   |  Description  |
// |----------|-------------|
// | \`primary\` | the principle call to action on the page |
// | \`secondary\` | secondary actions on each page   |
// | \`tertiary\` | simple actions on the page |
// | \`accent\` | very important actions and only once or twice on each page |
// | \`danger\` | a negative action (such as Delete) on the page |
// | \`danger--primary\` | a negative principle call to action (such as Delete) on the page |
// | \`ghost\` | in places where a regular button would draw too much attention and look similar to links |
// | \`inverse--primary\` | should be used on dark backgrounds |
// | \`inverse\` | should be used on dark backgrounds |
// | \`navigation\`| should be only used in the main navigation |
// `;

// ButtonKind.storyName = "Button kinds";
// ButtonKind.parameters = {
//   markdown: "hello",
//   layout: "centered",
//   docs: {
//     storyDescription: hello,
//   },
// };

// export const ButtonIcon = (args) => (
//   <>
//     <Button icon={AddCircle} kind="primary">
//       Button with icon
//     </Button>{" "}
//     <Button {...args} kind="primary" icon={AddCircle}></Button>{" "}
//     <Button {...args} iconReverse kind="secondary" icon={AddCircle}>
//       With iconReverse
//     </Button>{" "}
//     <Button href="#" icon={AddCircle} kind="ghost">
//       Button with icon link
//     </Button>
//   </>
// );

// const helloButtonIcon = `
// When words are not enough, icons can be used in buttons to better communicate what the button does. Icons should be
// always paired with text whenever possible.

// Use the \`icon\` and \`iconDescription\` prop to add an Icon to the Button. Refer to the [Icon](?selectedKind=Icon) documentation for usage.

// To show the Icon on the left side use the \`iconReverse\` prop.

// Using an Button with only an Icon while not beeing recommended can be achieved by leaving the Button text blank.
// `;

// const buttoniconsnippet = `
// import { Button } from '@wfp/react';
// import { AddCircleGlyph } from '@un/icons-react';

//   <>
//   <Button icon={AddCircleGlyph} kind="primary">
//     Button with icon
//   </Button>{' '}
//   <Button kind="tertiary" icon={AddCircleGlyph}></Button>{' '}
//   <Button iconReverse kind="tertiary" icon={AddCircleGlyph}>
//     With iconReverse
//   </Button>
//   </>
// `;

// ButtonIcon.parameters = {
//   docs: {
//     storyDescription: helloButtonIcon,
//     source: {
//       code: buttoniconsnippet,
//     },
//   },
// };
// ButtonIcon.storyName = "Button with icon";

// export const Small = (args) => (
//   <>
//     <Button kind="primary" small>
//       Small button
//     </Button>{" "}
//     <Button {...args} icon={AddCircle} kind="secondary" small>
//       With icon
//     </Button>
//   </>
// );

// const helloButtonSmall = `
// Small buttons may be used when there is not enough space for a
// regular sized button. This issue is most found in tables. Small button should have three words
// or less.
// `;

// const smallbuttonsnippet = `
// import { Button } from '@wfp/react';
// import { AddCircleGlyph } from '@un/icons-react';

//   <>
//     <Button small>Small button</Button>{' '}
//     <Button icon={AddCircleGlyph} kind="tertiary" small>
//       With icon
//     </Button>
//   </>
// `;

// Small.storyName = "small Button";
// Small.parameters = {
//   docs: {
//     storyDescription: helloButtonSmall,
//     source: {
//       code: smallbuttonsnippet,
//     },
//   },
// };

// export const Large = (args) => (
//   <>
//     <Button kind="primary" large>
//       Large button
//     </Button>
//   </>
// );

// const helloButtonLarge = `
// Small buttons may be used when there is not enough space for a
// regular sized button. This issue is most found in tables. Small button should have three words
// or less.
// `;

// const largebuttonsnippet = `
// import { Button } from '@wfp/react';
// import { AddCircleGlyph } from '@un/icons-react';

//   <>
//     <Button large>Large button</Button>{' '}
//     <Button {...args} icon={AddCircleGlyph} kind="tertiary" large>
//       With icon
//     </Button>
//   </>
// `;

// Large.storyName = "large Button";
// Large.parameters = {
//   docs: {
//     storyDescription: helloButtonLarge,
//     source: {
//       code: largebuttonsnippet,
//     },
//   },
// };

// const Icon = () => <div>Hello</div>;

// export const CustomIcon = () => (
//   <>
//     <Button large icon={Icon}>
//       Large button
//     </Button>
//   </>
// );

// // const CustomIconText = `
// // Small buttons may be used when there is not enough space for a
// // regular sized button. This issue is most found in tables. Small button should have three words
// // or less.
// // `;

// CustomIcon.storyName = "Custom Button";
