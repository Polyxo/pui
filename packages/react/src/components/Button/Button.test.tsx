import "@testing-library/jest-dom";
//import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Button from "./Button";

describe("Button Component", () => {
  // Test basic rendering
  it("renders a button", () => {
    render(<Button>Click me</Button>);
    expect(screen.getByRole("button")).toHaveTextContent("Click me");
  });

  // Test small and large variants
  it("renders a small button", () => {
    render(<Button small>Small Button</Button>);
    expect(screen.getByRole("button")).toHaveClass("wfp--btn--sm");
  });

  it("renders a large button", () => {
    render(<Button large>Large Button</Button>);
    expect(screen.getByRole("button")).toHaveClass("wfp--btn--lg");
  });

  // Test href prop
  it("renders an anchor when href is provided", () => {
    render(<Button href="https://example.com">Link</Button>);
    expect(screen.getByRole("button")).toHaveAttribute(
      "href",
      "https://example.com"
    );
  });

  // Test onClick event
  it("handles onClick event", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Clickable</Button>);
    fireEvent.click(screen.getByText("Clickable"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
  // Test if the button is disabled when the 'disabled' prop is passed
  it("renders a disabled button", () => {
    render(<Button disabled>Disabled Button</Button>);
    expect(screen.getByRole("button")).toBeDisabled();
  });

  // Test custom class name
  it("applies custom class names", () => {
    const customClass = "my-custom-class";
    render(<Button className={customClass}>Custom Class Button</Button>);
    expect(screen.getByRole("button")).toHaveClass(customClass);
  });

  // Test different kinds of buttons
  it("applies the primary kind class", () => {
    render(<Button kind="primary">Primary Button</Button>);
    expect(screen.getByRole("button")).toHaveClass("wfp--btn--primary");
  });

  // Test tabIndex prop
  it("sets tabIndex when provided", () => {
    const tabIndex = 3;
    render(<Button tabIndex={tabIndex}>Tab Indexed Button</Button>);
    expect(screen.getByRole("button")).toHaveAttribute(
      "tabIndex",
      tabIndex.toString()
    );
  });

  // Test large and small props together
  it("does not apply both large and small classes together", () => {
    render(
      <Button large small>
        Conflicting Sizes Button
      </Button>
    );
    const button = screen.getByRole("button");
    expect(button).not.toHaveClass("no-prefix--btn--lg");
    expect(button).not.toHaveClass("no-prefix--btn--sm");
  });

  // Test with children and icon
  it("renders children and icon together", () => {
    const MockIcon = () => <svg data-testid="mock-icon"></svg>;
    render(<Button icon={<MockIcon />}>Button with Icon</Button>);
    expect(screen.getByTestId("mock-icon")).toBeInTheDocument();
    expect(screen.getByRole("button")).toHaveTextContent("Button with Icon");
  });

  // Snapshot test
  it("matches snapshot", () => {
    const { asFragment } = render(<Button>Snapshot Button</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  // Test mouse events
  it("handles mouse events", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Clickable Button</Button>);
    fireEvent.click(screen.getByRole("button"));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});

// Mock icon as a simple SVG
const MockIcon = () => <svg data-testid="mock-icon"></svg>;

describe("Button Component with Icon", () => {
  // Test icon rendering when icon is a React node
  it("renders a button with an icon as a React node", () => {
    render(<Button icon={<MockIcon />}>Button with Icon</Button>);
    const icon = screen.getByTestId("mock-icon");
    expect(icon).toBeInTheDocument();
  });

  // Test icon rendering when icon is a component type
  it("renders a button with an icon as a component type", () => {
    render(<Button icon={MockIcon}>Button with Component Icon</Button>);
    const icon = screen.getByTestId("mock-icon");
    expect(icon).toBeInTheDocument();
  });

  // Test iconReverse prop
  it("renders with icon reverse", () => {
    const MockIcon = () => <svg data-testid="mock-icon"></svg>;
    render(
      <Button icon={MockIcon} iconReverse>
        Icon Reverse Button
      </Button>
    );

    const button = screen.getByRole("button");
    const icon = screen.getByTestId("mock-icon");
    expect(button.firstChild).toBe(icon);
  });
});
