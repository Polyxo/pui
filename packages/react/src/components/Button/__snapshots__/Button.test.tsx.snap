// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="wfp--btn wfp--btn--primary"
    type="button"
  >
    Snapshot Button
  </button>
</DocumentFragment>
`;
